Add search suggestions with autocomplete to home page

Ch4s3 opened the pull request:
This PR adds a LiveView-based autocomplete search on the homepage that suggests packages as users type.

#### User interface:
- Replaces the static search form on the homepage with a LiveView, `SearchSuggestionsComponent`
- Shows up to 8 suggestions in the dropdown as a users types
- Supports keyboard navigation with arrow keys and Enter to navigate to the package page
- Adds `Cmd+K`/`Ctrl+K` keyboard shortcut to focus the search input

#### Component structure:
- Implements `SearchSuggestionsLive` with debounced search and keyboard event handling
- Suggestions dropdown built as a functional component `SearchSuggestionsComponent`

#### Search implementation:
- Adds `Packages.suggest/3` with weighted ranking: exact match, prefix match, substring match, full-text description search, and recent downloads
- Uses [PostgreSQL trigram](https://www.postgresql.org/docs/current/pgtrgm.html) for name matching and GIN indexes for full-text search
- Returns highlighted descriptions with matching terms wrapped in `<strong>` tags

ericmj commented on lib/hexpm/repository/packages.ex:
```diff
@@ -122,4 +123,266 @@ defmodule Hexpm.Repository.Packages do
+      iex> Packages.suggest(repository, "")
+      []
+  """
+  def suggest(repository, term, limit \\ 8)
```
Could we incorporate this for the normal package search also?

Ch4s3 commented on assets/js/app.js:
```diff
@@ -92,6 +92,26 @@ export default class App {
       })
     }
+
+    // Global shortcut: Cmd+K / Ctrl+K focuses the search bar
```
I'd say it's increasingly common, [Tailwind](https://tailwindcss.com/) does it and its an Algolia feature. I can take or leave it, this is a pretty common design idiom for it 
<img width="74" height="44" alt="image" src="https://github.com/user-attachments/assets/6bfa65ad-4f87-4bfc-8042-3f02d9cf2546" />

Ch4s3 committed:
Add LiveView-powered search autocomplete
Replaces the static search forms on the navbar and home page hero with
a LiveView autocomplete component that suggests packages as the user types.
Search implementation (Packages.suggest/3):
- Weighted ranking: exact name match (3.0) > prefix (2.0) > substring (1.0),
  with recent-download scaling via ln(1 + downloads) * 0.2 and full-text
  description relevance via ts_rank_cd * 0.4
- Name matching uses LIKE against the existing btree index; description
  matching uses to_tsvector / to_tsquery against the GIN index on meta
- ts_headline extracts a 5–15 word description snippet with matched terms
  wrapped in <strong> for display
- Search term is capped at 100 chars; individual tsquery words at 50 chars
- Understands "repo/package" input format, searches only the package part
- Returns up to 8 results by default, configurable via the limit parameter
UI (SearchSuggestionsLive + SearchSuggestionsComponent):
- Three session variants: "home" (white, with Search button), "home-mobile"
  (same styles, distinct element IDs for the mobile hero), and "nav" (dark,
  no button, visible "/" shortcut badge)
- Debounced at 100 ms; closes on click-away
- Keyboard navigation: ArrowDown/Up move the highlight (wrapping), Escape
  closes the dropdown; active is nil by default so Enter without an explicit
  arrow-key selection falls through to a full-text search instead of
  hijacking the first suggestion
- phx-keydown on the input rather than phx-window-keydown on the container,
  so arrow keys don't interfere with other inputs on the page
- ARIA combobox / listbox / activedescendant attributes throughout
SearchShortcut JS hook:
- Extended from "/" only to also handle Cmd+K / Ctrl+K
- Skips inputs where offsetParent === null (hidden on mobile) so the
  shortcut always lands on the visible input
- stopImmediatePropagation prevents double-firing when multiple hook
  instances are mounted simultaneously
Security:
- ts_headline output is split on <strong> / </strong> tags and the
  surrounding text is HTML-escaped before being marked safe, preventing
  XSS from user-supplied package descriptions
- Name highlighting HTML-escapes both the name and the search term before
  regex substitution, then wraps matched text in <mark> with a case-
  insensitive pattern so original letter casing is preserved in output
Other:
- mark { background-color: transparent; color: inherit } reset in Tailwind
  base layer so browser default yellow highlight doesn't override styles
- lazy_html added as a test-only dependency (required by Phoenix.LiveViewTest)
- conn threaded through header → desktop_nav → search_form in navbar so
  live_render has access to the Plug connection

Ch4s3 committed:
Merge upstream/main into improve-search

Resolves conflicts introduced by the UI redesign on upstream:
- Delete old .eex templates and SCSS files removed by upstream
- Take upstream versions for all non-search files (accounts, auth,
  router, migrations, config, tests)
- Re-apply search feature changes on top of the redesigned files:
  - navbar.ex: conn attr threaded to header/desktop_nav/search_form;
    search_form replaced with live_render(SearchSuggestionsLive)
  - app.html.heex: conn={@conn} passed to <.header>
  - index.html.heex: LiveView search blocks added to mobile and
    desktop hero sections
  - tailwind.css: mark reset added to @layer base
  - search_shortcut.js: Cmd+K, offsetParent, stopImmediatePropagation
- lazy_html added to mix.exs (test only); cc_precompiler, fine,
  and lazy_html lock entries resolved via mix deps.get

Ch4s3 committed:
Add SearchSuggestionsLive tests and fix related bugs

Tests (14 total, async: false with shared sandbox):
- home variant: renders, suggests, hides, ArrowDown highlight, submit
  without selection → text search, ArrowDown+submit → package page,
  no-match submit, Escape, close event
- home-mobile variant: distinct element IDs from home
- nav variant: renders, suggests, submit without selection → text
  search, ArrowDown+submit → package page

Bugs fixed while writing tests:
- active initializes to nil (was 0) so Enter without an explicit
  ArrowDown falls through to text search instead of always navigating
  to the first suggestion
- ArrowDown/Up guard for count==0 and handle nil active (nil→0 for
  Down, nil→count-1 for Up)
- Escape handler added: closes dropdown and resets active to nil
- phx-keydown on input instead of phx-window-keydown on container
- phx-hook="SearchShortcut" added to inputs; two render clauses
  (nav vs home) with Tailwind classes instead of Bootstrap
- input_id/1 now returns distinct IDs for each variant:
  home→"home-search-input", home-mobile→"home-mobile-search-input",
  nav→"nav-search-input"
- listbox_id/1 adds "home-mobile"→"home-mobile-suggest-list" so
  home-mobile has a distinct ARIA listbox from home
- aria-selected now renders "true"/"false" strings (was an Elixir
  boolean which Phoenix renders as a bare attribute or omits it)
- Remove stray IO.inspect debug call in Packages.suggest/3

Ch4s3 committed:
Fix CI: update packages_suggest_test for {:safe, _} name_html type

Ch4s3 committed:
Fix suggest ranking, highlight_name, and safe_description_html

Scoring:
- Replace three cumulative CASE statements with a single exclusive CASE
  so packages land in exactly one tier: exact=3.0, prefix=2.0,
  substring=1.0. The old form added all matching tiers (exact got 6.0),
  producing correct ordering but values inconsistent with the comments.

highlight_name:
- Use a case-insensitive Regex instead of String.replace so "Ecto"
  highlights the "Ecto" portion of "EctoSQL" and vice versa.
- Replace fn preserves the original casing of the matched characters
  from the package name rather than substituting the typed term.
- Use <mark> instead of <strong> to match the mark CSS reset in
  tailwind.css.

safe_description_html:
- Replace Phoenix.HTML.raw(description) with an explicit split on the
  <strong>…</strong> delimiters injected by ts_headline. Text segments
  outside those delimiters are HTML-escaped before reassembly, preventing
  XSS from user-supplied package descriptions that contain HTML.

Tests:
- name_html wraps matched portion in a mark element
- name_html match is case-insensitive and preserves original casing
- description_html does not pass raw HTML tags from descriptions

Ch4s3 commented:
@ericmj it took me a bit to get back to this, but I think it should be ready for review now.

ericmj commented:
I am afraid this PR will conflict with https://github.com/hexpm/hexpm/pull/1496. Do you think this can be incorporated into 1496?

I think the extra search bar on the front page is unnecessary when we have one in the navbar.

I also see some rendering issues:
<img width="1624" height="1060" alt="Screenshot 2026-04-28 at 16 28 28" src="https://github.com/user-attachments/assets/e1dc1e9f-c1e6-4ab3-8d69-589239fc40ae" />
<img width="1624" height="1060" alt="Screenshot 2026-04-28 at 16 28 21" src="https://github.com/user-attachments/assets/83d101b3-ee99-4c7b-960c-a28291571fb3" />

Ch4s3 commented:
@ericmj I can definitely incorporate it. Should I point this PR to your branch?


I just saw that rendering issue too. Should be just a css thing.

Ch4s3 commented:
@ericmj I'm leaving this PR open for the moment if you don't mind but just opened https://github.com/hexpm/hexpm/pull/1531 pointing at your branch. Give me a bit to get it in shape.

Ch4s3 commented:
I'll probably close this in favor of #1531

Ch4s3 commented:
> I'll probably close this in favor of #1531
